### PR TITLE
Verify Email upon signup

### PR DESCRIPTION
**Pull request by riley**
## What does this PR do?

- adds a setting in the admin to enable email verification for new users
- email is sent to new users upon sign up
- clicking the link will confirm your account
- if the user got distracted upon signing up the first time, or lost the email, they can request a new one after trying to sign in with the right credentials

## How do I test this PR?

- turn on Require Email Confirmation in the admin
- create a new user
- click the confirm link in your email
- if you try to request an email again if you're already verified, it should bounce you
- try to log in again. it should work.

### CODE REVIEW

**Comment by davidgljay** [comment]
After verifying my e-mail I'm redirected to `http://localhost:3000/admin/undefined`. Is this intentional?

File: client/coral-sign-in/containers/SignInContainer.js [hunk 1]
```diff
@@ -71,6 +75,17 @@ class SignInContainer extends Component {
     });
   }
 
+  handleChangeEmail(e) {
+    const {value} = e.target;
+    console.log('handleChangeEmail', value);
```
Comment (davidgljay) [inline client/coral-sign-in/containers/SignInContainer.js]: This should be removed.

File: client/coral-sign-in/containers/SignInContainer.js [hunk 2]
```diff
@@ -71,6 +75,17 @@ class SignInContainer extends Component {
...
+    this.setState({emailToBeResent: value});
+  }
+
+  handleResendConfirmation(e) {
+    e.preventDefault();
+    this.props.requestConfirmEmail(this.state.emailToBeResent);
```
Comment (davidgljay) [inline client/coral-sign-in/containers/SignInContainer.js]: This doesn't seem to give the user a confirmation or close the dialog. It should probably do one or both.

**Review by davidgljay — CHANGES_REQUESTED** [review]
Remove a console.log and address issue with resent confirmation e-mail dialog not closing.

**Review by wyattjoh — COMMENTED** [review]
Should handle the state where the referer isn't defined or isn't a url first, possibly create a validateReferer method to the SettingsService and call out to it?

File: services/users.js [hunk 3]
```diff
@@ -538,40 +538,51 @@ module.exports = class UsersService {
    * @param  {String} email The email that we are needing to get confirmed.
    * @return {Promise}
    */
-  static createEmailConfirmToken(userID, email) {
+  static createEmailConfirmToken(userID = null, email, referer) {
```
Comment (wyattjoh) [inline services/users.js]: Should default to something here if referer isn't provided.

File: routes/api/users/index.js [hunk 4]
```diff
@@ -113,10 +116,12 @@ const SendEmailConfirmation = (app, userID, email) => UsersService
     });
   });
 
+// create a local user.
 router.post('/', (req, res, next) => {
   const {
     email,
     password,
+    redirectUri,
```
Comment (wyattjoh) [inline routes/api/users/index.js]: Should source this from the `Referrer` header, not a payload option.

**Review by wyattjoh — CHANGES_REQUESTED** [review]

File: routes/api/users/index.js (hunk 4 above)
Comment (riley) [reply routes/api/users/index.js]: whoops, good catch. I removed that from somewhere else when I was moving things around and forgot to re-add it here.

**Review by wyattjoh — APPROVED** [review]

**Review by davidgljay — APPROVED** [review]

**Pull request merged**